lib/membrane_mp4/muxer/cmaf/track_samples_queue.ex

```diff
@@ -0,0 +1,280 @@
+defmodule Membrane.MP4.Muxer.CMAF.TrackSamplesQueue do
+  @moduledoc false
+
+  defstruct collectable?: false,
+            track_with_keyframes?: false,
+            collected_duration: 0,
+            target_samples: [],
+            to_collect_duration: 0,
+            excess_samples: []
+
+  @type state_t :: :collecting | :to_collect
+
+  @type t :: %__MODULE__{
+          collectable?: boolean(),
+          track_with_keyframes?: boolean(),
+          collected_duration: non_neg_integer(),
+          target_samples: list(Membrane.Buffer.t()),
+          excess_samples: list(Membrane.Buffer.t())
+        }
+
+  @doc """
+  Force pushes the sample into the queue ignoring any timestamps.
+  """
+  @spec force_push(t(), Membrane.Buffer.t()) :: t()
+  def force_push(%__MODULE__{collectable?: false} = queue, sample) do
+    %__MODULE__{
+      queue
+      | target_samples: [sample | queue.target_samples],
+        collected_duration: queue.collected_duration + sample.metadata.duration
+    }
+  end
+
+  @doc """
+  Pushes sample into the queue based on the given timestamp and sample's dts.
+
+  If the sample has a dts lower than the desired timestamp then the sample will land in target samples group,
+  otherwise it means that the required samples are already collected and the sample will land in excess samples group
+  and queue will become collectable.
+  """
+  @spec push(t(), Membrane.Buffer.t(), Membrane.Time.t()) :: t()
+  def push(%__MODULE__{collectable?: false} = queue, sample, timestamp) do
+    dts = Ratio.to_float(sample.dts)
+
+    if dts < timestamp do
+      %__MODULE__{
+        queue
+        | collected_duration: queue.collected_duration + sample.metadata.duration,
+          target_samples: [sample | queue.target_samples]
+      }
+    else
+      %__MODULE__{
+        queue
+        | collectable?: true,
+          target_samples: Enum.reverse(queue.target_samples),
+          excess_samples: [sample]
+      }
+    end
```

Comment:
Shouldn't we switch to collectable when `dts + sample.metadata.duration < timestamp`? I still find it quite confusing that we're operating on both timestamps and durations. If we operated e.g. only on durations, we could pass target duration as an option to the queue and for example switch to collectable if `collected_duration >= target_duration`.

Reply:
> Shouldn't we switch to collectable when dts + sample.metadata.duration < timestamp? 

It probably depends if we want eventually to land with a segment that is a little bit shorter (if we account for the sample's duration in comparison) or a little bit longer (the dts + duration spans a little bit into the next segment). For me at least it doesn't make a difference (unless it matters somewhere else).

And you probably meant to switch to collectable when `dts + smaple.metadata.duration >= timestamp` (the else statement)?

Reply:
> And you probably meant to switch to collectable when dts + smaple.metadata.duration >= timestamp (the else statement)?

Yes that's what I meant

Reply:
Refactored the queue to operate on durations instead.